extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java

```diff
@@ -298,4 +336,119 @@ private ResultHandle generateSort(Integer sortParameterIndex, MethodCreator meth
         }
         return sort;
     }
+
+    // Make sure the return type is referencing a class we have indexed 
+    // somewhere along the way (e.g. non-entity return types in @Query methods)
+    // Unless it is some kind of collection containing multiple types, 
+    // return the type used in the query result.
+    private Type verifyQueryResultType(Type t) {
+        if (t.kind() == Kind.ARRAY) {
+            return verifyQueryResultType(t.asArrayType().component());
+        } else if (t.kind() == Kind.PARAMETERIZED_TYPE) {
+            List<Type> list = t.asParameterizedType().arguments();
+            if (list.size() == 1) {
+                return verifyQueryResultType(list.get(0));
+            } else {
+                for (Type x : list) {
+                    verifyQueryResultType(x);
+                }
+                return t;
+            }
+        } else if (!DotNames.OBJECT.equals(t.name())) {
+            ClassInfo typeClassInfo = index.getClassByName(t.name());
+            if (typeClassInfo == null) {
+                throw new IllegalStateException(t.name() + " was not part of the Quarkus index");
+            }
+        }
+        return t;
+    }
+
+    // Verify the implementation class exists (generate it if necessary)
+    private DotName verifyGeneratedImplementationClass(ClassInfo interfaceInfo, List<String> fieldNames) {
+        DotName interfaceName = interfaceInfo.name();
+        DotName customImplName = customReturnTypes.get(interfaceName);
+
+        Map<String, FieldDescriptor> fields = new HashMap<>(fieldNames.size());
+        if (customImplName == null) {
+            customImplName = GenerationUtil.createImplDotName(interfaceName);
+            try (ClassCreator implClassCreator = ClassCreator.builder().classOutput(nonBeansClassOutput)
+                    .interfaces(interfaceName.toString()).className(customImplName.toString())
+                    .build()) {
+
+                List<MethodInfo> methods = interfaceInfo.methods();
+
+                // Work with methods in declared order
+                for (int i = 0; i < methods.size(); i++) {
```

Comment:
for each loop perhaps would be nicer?

Reply:
I can go to a foreach loop now, yes. Remember that thing where I was dependent on order? That's from then. ;) I'll fix